src/Features/LanguageServer/Protocol/Handler/RequestContext.cs

```diff
@@ -44,22 +45,31 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler;
 
     private readonly ILspServices _lspServices;
 
+    /// <summary>
+    /// Provides backing storage for the LSP workspace used by this RequestContext instance, allowing it to be cleared
+    /// on demand from all copies that may exist of this value type.
+    /// </summary>
+    /// <remarks>
+    /// This field is only initialized for handlers that request solution context.
+    /// </remarks>
+    private readonly StrongBox<(Workspace Workspace, Solution Solution, Document? Document)>? _lspSolution;
+
     /// <summary>
     /// The workspace this request is for, if applicable.  This will be present if <see cref="Document"/> is
     /// present.  It will be <see langword="null"/> if <c>requiresLSPSolution</c> is false.
     /// </summary>
-    public readonly Workspace? Workspace;
+    public Workspace? Workspace => _lspSolution?.Value.Workspace;
```

Comment:
Can we make the exception a bit clearer if you use this after ClearSolutionContext is called? I assume you're doing the StrongBox to differentiate "this never had it" and "this was cleared", but that's not really confirmed anywhere.

(and frankly, I'm not sure why these wouldn't just throw if requiresLSPSolution is set, since I'd imagine that's a developer oops, but maybe I'm missing some context here)

Reply:
> I assume you're doing the StrongBox to differentiate "this never had it" and "this was cleared"

No, the `StrongBox<>` is required because every time `RequestContext` is passed as an argument, a new copy is made. We need a common storage location for the references to Workspace and Solution that applies to all copies equally.

Reply:
➡️ Added code to throw for access to these properties if the solution context was initially available _and_ has since been cleared.